src/types/query.d.ts

```diff
@@ -1,46 +1,172 @@
+
+export type UpdateParams<T extends Updatable> = unknown &
+  T extends Expr.DocumentRef<infer U>
+  ? {
+      data?: Partial<U>
+      credentials?: object
+      delegates?: object
+    }
+  : T extends Expr.Ref<infer U>
+  ? Omit<Partial<U>, 'ref' | 'ts'>
+  : never
+
+export type CreateCollectionParams<Meta extends object = any> = {
+  name: string
+  data?: Meta
+  permissions?: object
+  history_days?: number
+  ttl_days?: number
+}
+
+export type CreateIndexParams<Meta extends object = any> = {
+  name: string
+  source: Expr.CollectionRef<any> | any[]
+  terms?: any[]
+  values?: any[]
+  unique?: boolean
+  serialized?: boolean
+  permissions?: object
+  data?: Meta
+}
+
+export type PaginateParams = {
+  ts?: number | Expr.Time
+  size?: number
+  before?: string
+  after?: string
+  events?: boolean
+  sources?: boolean
+}
+
+export type FunctionParams<Meta extends object = any> = {
+  name: string
+  body: object
+  data?: Meta
+  role?: any
+}
 
 export module query {
-  export function Ref(ref: ExprArg, id?: ExprArg): Expr
+  export function Ref<T extends object>(
+    ref: Expr.CollectionRef<T>,
+    id?: ExprVal<number | string>
+  ): Expr.DocumentRef<T>
+
+  // TODO
   export function Bytes(bytes: ExprArg | ArrayBuffer | Uint8Array): Expr
   export function Abort(msg: ExprArg): Expr
   export function At(timestamp: ExprArg, expr: ExprArg): Expr
-  export function Let(vars: ExprArg, in_expr: ExprArg): Expr
+
+  export function Let<T>(vars: ExprVal<object>, in_expr: Expr<T>): Expr<T>
+
+  // TODO
   export function Var(varName: ExprArg): Expr
-  export function If(
-    condition: ExprArg,
-    then: ExprArg | null,
-    _else: ExprArg | null
-  ): Expr
-  export function Do(...args: ExprArg[]): Expr
+
+  export function If<T, U>(
+    condition: ExprVal<boolean>,
+    then: ExprVal<T>,
+    _else: ExprVal<U>
+  ): ToExpr<T | U>
+
+  export function If<T>(
+    condition: ExprVal<boolean>,
+    then: ExprVal<T>,
+    _else: ExprVal<T>
+  ): ToExpr<T>
+
+  export function Do<T>(arg1: ExprVal<T>): ToExpr<T>
+  export function Do<T>(arg1: ExprVal, arg2: ExprVal<T>): ToExpr<T>
+  export function Do<T>(
+    arg1: ExprVal,
+    arg2: ExprVal,
+    arg3: ExprVal<T>
+  ): ToExpr<T>
+  export function Do<T>(
+    arg1: ExprVal,
+    arg2: ExprVal,
+    arg3: ExprVal,
+    arg4: ExprVal<T>
+  ): ToExpr<T>
+  export function Do<T>(
+    arg1: ExprVal,
+    arg2: ExprVal,
+    arg3: ExprVal,
+    arg4: ExprVal,
+    arg5: ExprVal<T>
+  ): ToExpr<T>
+  export function Do<T>(...args: ExprVal[]): ToExpr<T>
+
+  // TODO
   export function Object(fields: ExprArg): Expr
-  export function Lambda(f: Lambda): Expr
-  export function Lambda(var_name: ExprArg, expr: ExprArg): Expr
-  export function Call(ref: ExprArg, ...args: ExprArg[]): Expr
+
+  export function Lambda<In extends any[], Out>(
+    f: (...args: In) => Out
+  ): Expr<Lambda<In, Out>>
+
+  export function Lambda<In extends any[], Out>(
+    var_name: ExprArg<string>,
```

Comment:
In that form, `var_name` is either a single string or an array of strings, never an expression